src/libpmem2/badblocks_ndctl.c

```diff
@@ -806,3 +807,220 @@ badblocks_clear_all(const char *file)
 
 	return ret;
 }
+
+int
+pmem2_badblock_iterator_new(const struct pmem2_source *src,
+		struct pmem2_badblock_iterator **pbb)
+{
+	ASSERTne(pbb, NULL);
+
+	struct ndctl_ctx *ctx;
+	struct ndctl_region *region;
+	struct ndctl_namespace *ndns;
+	os_stat_t st;
+	int ret = -1;
+
+	if (ndctl_new(&ctx)) {
+		ERR("!ndctl_new");
+		return PMEM2_E_ERRNO;
+	}
+
+	if (os_fstat(src->fd, &st)) {
+		ERR("!fstat %i", src->fd);
+		ret = PMEM2_E_ERRNO;
+		goto error_ndctl_unref;
+	}
+
+	int rv = ndctl_region_namespace(ctx, &st, &region, &ndns);
+	if (rv) {
+		LOG(1, "getting region and namespace failed");
+		ret = rv;
+		goto error_ndctl_unref;
+	}
+
+	if (ndns == NULL) {
+		ERR("cannot get namespace");
+		goto error_ndctl_unref;
+	}
+
+	/*
+	 * Only the new NDCTL versions have the namespace badblock iterator.
+	 * When compiled with older versions, the library needs to rely
+	 * on the old region interface.
+	 */
+	if (ndctl_namespace_get_mode(ndns) == NDCTL_NS_MODE_FSDAX) {
+		struct pmem2_badblock_iterator *bb =
+			Zalloc(sizeof(struct pmem2_badblock_iterator));
+		if (bb == NULL) {
+			ERR("!Zalloc");
+			ret = PMEM2_E_ERRNO;
+			goto error_ndctl_unref;
+		}
```

Comment:
I think it makes sense to add pmem2_zalloc, just like we have pmem2_malloc.

Reply:
Done